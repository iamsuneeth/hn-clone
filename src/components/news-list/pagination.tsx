/** @jsx jsx */
import { css, jsx } from "@emotion/core";
import { colors } from "../../theme/constants";
import { Link } from "react-router-dom";

const seperator = css`
  color: ${colors.primary};
  padding: 0 0.2rem;
`;

const paginationContainer = css`
  display: flex;
  justify-content: flex-end;
  padding: 0.2rem;
  @media screen and (max-width: 767px) {
    justify-content: center;
  }
  a {
    color: ${colors.primary};
    text-decoration: none;
    &:visited,
    &:active,
    &:focus,
    &:link {
      color: ${colors.primary};
    }
  }
`;

interface PaginationProps {
  hasPrevPage: boolean;
  hasNextPage: boolean;
  currentPage: number;
}

export const Pagination = ({
  hasPrevPage,
  hasNextPage,
  currentPage,
}: PaginationProps) => {
  return (
    <div css={paginationContainer}>
      {hasPrevPage && (
        <Link to={`/${currentPage === 1 ? "" : currentPage - 1}`}>
          Previous
        </Link>
      )}
      {hasNextPage && hasPrevPage && <span css={seperator}>|</span>}
      {hasNextPage && <Link to={`/${currentPage + 1}`}>Next</Link>}
    </div>
  );
};
