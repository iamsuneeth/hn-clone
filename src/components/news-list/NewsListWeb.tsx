/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React from "react";
import { css, jsx } from "@emotion/core";
import TimeAgo from "react-timeago";
import { Caret } from "../elements/icons/caret";
import { IconButton } from "../elements/buttons/IconButton";
import { colors } from "../../theme/constants";
import { newsListProps } from "../home/NewsList";
import { Pagination } from "./pagination";
import { Button } from "../elements/buttons/Button";

const commonRowStyle = css`
  display: flex;
  align-items: center;
  & > div {
    padding: 0.1rem 0.5rem;
    min-width: 5rem;
    text-align: center;
    &:last-child {
      flex: 1;
      text-align: left;
    }
  }
`;

const tableHeaderRowStyle = css`
  ${commonRowStyle};
  background-color: ${colors.primary};
  color: ${colors.textWhite};
`;

const tableContainerStyle = css`
  background-color: ${colors.background};
  margin: 0.5rem 0;
`;

const tableRowStyle = css`
  ${commonRowStyle}
  &:nth-of-type(2n) {
    background-color: ${colors.alternateBackground};
  }
`;

const secondaryStyle = css`
  color: ${colors.textColorSecondary};
  font-size: 90%;
`;

const newsDetailStyle = css`
  display: flex;
  flex-wrap: wrap;
  & > div {
    & > span {
      padding: 0 0.1rem;
    }
  }
  & > span {
    padding: 0 0.1rem;
  }
  a {
    text-decoration: none;
    color: ${colors.textColorSecondary};
    :hover {
      text-decoration: underline;
      text-decoration-color: inherit;
    }
  }

  @media screen and (max-width: 1079px) {
    flex-direction: column;
  }
`;

const hideButtonStyle = css`
  &:hover {
    text-decoration: underline;
  }
`;

export const NewsListWeb = ({
  items,
  hasNextPage,
  hasPrevPage,
  currentPage,
  hide,
  upVote,
}: newsListProps) => {
  return (
    <div
      css={tableContainerStyle}
      role="table"
      aria-label="newslist"
      data-id="web"
    >
      <div role="rowgroup">
        <div role="row" css={tableHeaderRowStyle}>
          <div role="columnheader">Comments</div>
          <div role="columnheader">Vote Count</div>
          <div role="columnheader">Upvote</div>
          <div role="columnheader">News Details</div>
        </div>
      </div>
      <div role="rowgroup">
        {items.map((item) => {
          const url = item.url ? new URL(item.url) : null;
          return (
            <div role="row" css={tableRowStyle} key={item.objectID}>
              <div role="cell">{item.num_comments}</div>
              <div role="cell">{item.points}</div>
              <div role="cell">
                <IconButton
                  aria-label="upvote"
                  title="upVoteButton"
                  icon={<Caret />}
                  onClick={() => upVote(item.objectID)}
                />
              </div>
              <div role="cell" css={newsDetailStyle}>
                <div>
                  <span>{item.title}</span>
                  {url && (
                    <a href={item.url}>
                      <span css={secondaryStyle}>{`(${url.host})`}</span>
                    </a>
                  )}
                </div>
                <div>
                  <span css={secondaryStyle}>by</span>
                  <span>{item.author}</span>
                  <span css={secondaryStyle}>
                    <TimeAgo date={item.created_at} />
                  </span>
                  <span>
                    [
                    <Button
                      aria-label="hide"
                      css={hideButtonStyle}
                      onClick={() => hide(item.objectID)}
                      title="hideButton"
                    >
                      Hide
                    </Button>
                    ]
                  </span>
                </div>
              </div>
            </div>
          );
        })}
      </div>
      <Pagination
        hasPrevPage={hasPrevPage}
        hasNextPage={hasNextPage}
        currentPage={currentPage}
      />
    </div>
  );
};
