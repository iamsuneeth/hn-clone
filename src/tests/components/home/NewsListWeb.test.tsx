import React from "react";
import { render } from "@testing-library/react";
import { newsListTestData } from "../../data/newsList";
import { NewsListWeb } from "../../../components/news-list/NewsListWeb";
import { BrowserRouter as Router } from "react-router-dom";

describe("Tests for NewsList component", () => {
  test("should render with web version of news", () => {
    const upVote = jest.fn();
    const hide = jest.fn();

    const { getAllByLabelText, container } = render(
      <Router>
        <NewsListWeb
          items={newsListTestData.hits}
          currentPage={0}
          hasNextPage
          hasPrevPage
          upVote={upVote}
          hide={hide}
        />
      </Router>
    );

    expect(getAllByLabelText("upvote").length).toBeGreaterThan(0);
    expect(container).toMatchSnapshot();
  });
});
