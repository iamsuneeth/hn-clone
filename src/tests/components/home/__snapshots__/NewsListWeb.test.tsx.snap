// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for NewsList component should render with web version of news 1`] = `
<div>
  <div
    aria-label="newslist"
    class="css-jsmc8z-NewsListWeb"
    data-id="web"
    role="table"
  >
    <div
      role="rowgroup"
    >
      <div
        class="css-1hfr4j0-NewsListWeb"
        role="row"
      >
        <div
          role="columnheader"
        >
          Comments
        </div>
        <div
          role="columnheader"
        >
          Vote Count
        </div>
        <div
          role="columnheader"
        >
          Upvote
        </div>
        <div
          role="columnheader"
        >
          News Details
        </div>
      </div>
    </div>
    <div
      role="rowgroup"
    >
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          436
        </div>
        <div
          role="cell"
        >
          6015
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Stephen Hawking has died
            </span>
            <a
              href="http://www.bbc.com/news/uk-43396008"
            >
              <span
                class="css-n9etz9"
              >
                (www.bbc.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              Cogito
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2018-03-14T03:50:30.000Z"
                title="2018-03-14T03:50:30.000Z"
              >
                2 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          967
        </div>
        <div
          role="cell"
        >
          5771
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              A Message to Our Customers
            </span>
            <a
              href="http://www.apple.com/customer-letter/"
            >
              <span
                class="css-n9etz9"
              >
                (www.apple.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              epaga
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2016-02-17T08:38:37.000Z"
                title="2016-02-17T08:38:37.000Z"
              >
                4 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          376
        </div>
        <div
          role="cell"
        >
          4271
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Steve Jobs has passed away.
            </span>
            <a
              href="http://www.apple.com/stevejobs/"
            >
              <span
                class="css-n9etz9"
              >
                (www.apple.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              patricktomas
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2011-10-05T23:42:23.000Z"
                title="2011-10-05T23:42:23.000Z"
              >
                9 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          1014
        </div>
        <div
          role="cell"
        >
          4107
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Reflecting on one very, very strange year at Uber
            </span>
            <a
              href="https://www.susanjfowler.com/blog/2017/2/19/reflecting-on-one-very-strange-year-at-uber"
            >
              <span
                class="css-n9etz9"
              >
                (www.susanjfowler.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              grey-area
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2017-02-19T21:16:33.000Z"
                title="2017-02-19T21:16:33.000Z"
              >
                3 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          1096
        </div>
        <div
          role="cell"
        >
          3816
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Bye, Amazon
            </span>
            <a
              href="https://www.tbray.org/ongoing/When/202x/2020/04/29/Leaving-Amazon"
            >
              <span
                class="css-n9etz9"
              >
                (www.tbray.org)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              grey-area
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2020-05-04T08:43:35.000Z"
                title="2020-05-04T08:43:35.000Z"
              >
                2 months ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          969
        </div>
        <div
          role="cell"
        >
          3592
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Every Google result now looks like an ad
            </span>
            <a
              href="https://twitter.com/craigmod/status/1219644556003565568"
            >
              <span
                class="css-n9etz9"
              >
                (twitter.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              cmod
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2020-01-21T15:38:22.000Z"
                title="2020-01-21T15:38:22.000Z"
              >
                6 months ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          1397
        </div>
        <div
          role="cell"
        >
          3384
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              F.C.C. Repeals Net Neutrality Rules
            </span>
            <a
              href="https://www.nytimes.com/2017/12/14/technology/net-neutrality-repeal-vote.html"
            >
              <span
                class="css-n9etz9"
              >
                (www.nytimes.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              panny
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2017-12-14T18:13:35.000Z"
                title="2017-12-14T18:13:35.000Z"
              >
                3 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          83
        </div>
        <div
          role="cell"
        >
          3381
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Show HN: This up votes itself
            </span>
            <a
              href="http://news.ycombinator.com/vote?for=3742902&dir=up&whence=%6e%65%77%65%73%74"
            >
              <span
                class="css-n9etz9"
              >
                (news.ycombinator.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              olalonde
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2012-03-23T00:40:39.000Z"
                title="2012-03-23T00:40:39.000Z"
              >
                8 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          981
        </div>
        <div
          role="cell"
        >
          3287
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Switch from Chrome to Firefox
            </span>
            <a
              href="https://www.mozilla.org/en-US/firefox/switch/"
            >
              <span
                class="css-n9etz9"
              >
                (www.mozilla.org)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              WisNorCan
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2019-05-30T16:09:19.000Z"
                title="2019-05-30T16:09:19.000Z"
              >
                1 year ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          472
        </div>
        <div
          role="cell"
        >
          3270
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Ask HN: I'm a software engineer going blind, how should I prepare?
            </span>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              zachrip
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2020-04-19T21:33:46.000Z"
                title="2020-04-19T21:33:46.000Z"
              >
                3 months ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          992
        </div>
        <div
          role="cell"
        >
          3238
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Cloudflare Reverse Proxies Are Dumping Uninitialized Memory
            </span>
            <a
              href="https://bugs.chromium.org/p/project-zero/issues/detail?id=1139"
            >
              <span
                class="css-n9etz9"
              >
                (bugs.chromium.org)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              tptacek
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2017-02-23T23:05:08.000Z"
                title="2017-02-23T23:05:08.000Z"
              >
                3 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          2531
        </div>
        <div
          role="cell"
        >
          3125
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              UK votes to leave EU
            </span>
            <a
              href="http://www.bbc.co.uk/news/uk-politics-36615028"
            >
              <span
                class="css-n9etz9"
              >
                (www.bbc.co.uk)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              dmmalam
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2016-06-24T03:48:57.000Z"
                title="2016-06-24T03:48:57.000Z"
              >
                4 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          974
        </div>
        <div
          role="cell"
        >
          3086
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Tim Cook Speaks Up
            </span>
            <a
              href="http://www.businessweek.com/articles/2014-10-30/tim-cook-im-proud-to-be-gay"
            >
              <span
                class="css-n9etz9"
              >
                (www.businessweek.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              replicatorblog
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2014-10-30T11:12:23.000Z"
                title="2014-10-30T11:12:23.000Z"
              >
                6 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          485
        </div>
        <div
          role="cell"
        >
          3030
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Announcing the first SHA-1 collision
            </span>
            <a
              href="https://security.googleblog.com/2017/02/announcing-first-sha1-collision.html"
            >
              <span
                class="css-n9etz9"
              >
                (security.googleblog.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              pfg
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2017-02-23T13:01:08.000Z"
                title="2017-02-23T13:01:08.000Z"
              >
                3 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          435
        </div>
        <div
          role="cell"
        >
          3015
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              I Sell Onions on the Internet
            </span>
            <a
              href="https://www.deepsouthventures.com/i-sell-onions-on-the-internet/"
            >
              <span
                class="css-n9etz9"
              >
                (www.deepsouthventures.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              eightturn
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2019-04-23T13:00:24.000Z"
                title="2019-04-23T13:00:24.000Z"
              >
                1 year ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          813
        </div>
        <div
          role="cell"
        >
          3001
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              macOS High Sierra: Anyone can login as “root” with empty password
            </span>
            <a
              href="https://twitter.com/lemiorhan/status/935578694541770752"
            >
              <span
                class="css-n9etz9"
              >
                (twitter.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              vladikoff
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2017-11-28T19:41:10.000Z"
                title="2017-11-28T19:41:10.000Z"
              >
                3 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          872
        </div>
        <div
          role="cell"
        >
          2968
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              SpaceX’s Falcon Heavy successfully launches
            </span>
            <a
              href="https://techcrunch.com/2018/02/06/spacexs-historic-falcon-heavy-successfully-launches/?ncid=rss&utm_source=dlvr.it&utm_medium=twitter"
            >
              <span
                class="css-n9etz9"
              >
                (techcrunch.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              mpweiher
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2018-02-06T20:57:15.000Z"
                title="2018-02-06T20:57:15.000Z"
              >
                2 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          1338
        </div>
        <div
          role="cell"
        >
          2953
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Twitter Will Allow Employees to Work at Home Forever
            </span>
            <a
              href="https://www.buzzfeednews.com/article/alexkantrowitz/twitter-will-allow-employees-to-work-at-home-forever"
            >
              <span
                class="css-n9etz9"
              >
                (www.buzzfeednews.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              minimaxir
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2020-05-12T16:09:46.000Z"
                title="2020-05-12T16:09:46.000Z"
              >
                2 months ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          277
        </div>
        <div
          role="cell"
        >
          2901
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Paul Allen has died
            </span>
            <a
              href="https://www.cnbc.com/2018/10/15/microsoft-co-founder-paul-allen-dies-of-cancer-at-age-65.html"
            >
              <span
                class="css-n9etz9"
              >
                (www.cnbc.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              coloneltcb
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2018-10-15T22:00:14.000Z"
                title="2018-10-15T22:00:14.000Z"
              >
                2 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-94mjpp"
        role="row"
      >
        <div
          role="cell"
        >
          1437
        </div>
        <div
          role="cell"
        >
          2873
        </div>
        <div
          role="cell"
        >
          <button
            aria-label="upvote"
            class="css-zxyc8n-Button"
            title="upVoteButton"
          >
            <div
              class="css-1a64y86-Icon"
            >
              <div
                class="css-eng7t9-Caret"
              />
            </div>
          </button>
        </div>
        <div
          class="css-koq5ar"
          role="cell"
        >
          <div>
            <span>
              Microsoft Is Said to Have Agreed to Acquire GitHub
            </span>
            <a
              href="http://www.bloomberg.com/news/articles/2018-06-03/microsoft-is-said-to-have-agreed-to-acquire-coding-site-github?"
            >
              <span
                class="css-n9etz9"
              >
                (www.bloomberg.com)
              </span>
            </a>
          </div>
          <div>
            <span
              class="css-n9etz9"
            >
              by
            </span>
            <span>
              miguelrochefort
            </span>
            <span
              class="css-n9etz9"
            >
              <time
                datetime="2018-06-03T20:14:26.000Z"
                title="2018-06-03T20:14:26.000Z"
              >
                2 years ago
              </time>
            </span>
            <span>
              [
              <button
                aria-label="hide"
                class="css-1cgq1qy-Button"
                title="hideButton"
              >
                Hide
              </button>
              ]
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-gi2tou-Pagination"
    >
      <a
        href="/-1"
      >
        Previous
      </a>
      <span
        class="css-9btmpj-Pagination"
      >
        |
      </span>
      <a
        href="/1"
      >
        Next
      </a>
    </div>
  </div>
</div>
`;
